web/web.go

```diff
@@ -0,0 +1,694 @@
+	if err != nil {
+		http.Error(w, fmt.Sprintf("Schema Conversion Error : %v", err), http.StatusNotFound)
+		return
+	}
+	app.conv = nil
+	app.conv = conv
+	app.driver = dc.Driver
+	w.WriteHeader(http.StatusOK)
+	json.NewEncoder(w).Encode(conv)
+}
+
+// TODO: Add Index key statements.
+func getDDL(w http.ResponseWriter, r *http.Request) {
+	c := ddl.Config{Comments: true, ProtectIds: false}
+	var tables []string
+	for t := range app.conv.SpSchema {
+		tables = append(tables, t)
+	}
+	sort.Strings(tables)
+	ddl := make(map[string]string)
+	for _, t := range tables {
+		ddl[t] = app.conv.SpSchema[t].PrintCreateTable(c)
+	}
+	w.WriteHeader(http.StatusOK)
+	json.NewEncoder(w).Encode(ddl)
+}
+
+func getSummary(w http.ResponseWriter, r *http.Request) {
+	reports := internal.AnalyzeTables(app.conv, nil)
+	summary := make(map[string]string)
+	for _, t := range reports {
+		var body string
+		for _, x := range t.Body {
+			body = body + x.Heading + "\n"
+			for i, l := range x.Lines {
+				body = body + fmt.Sprintf("%d) %s.\n", i+1, l) + "\n"
+			}
+		}
+		summary[t.SrcTable] = body
+	}
+	w.WriteHeader(http.StatusOK)
+	json.NewEncoder(w).Encode(summary)
+}
+
+func getOverview(w http.ResponseWriter, r *http.Request) {
+	var buf bytes.Buffer
+	bufWriter := bufio.NewWriter(&buf)
+	internal.GenerateReport(app.driver, app.conv, bufWriter, nil, false, false)
+	bufWriter.Flush()
+	overview := buf.String()
+	w.WriteHeader(http.StatusOK)
+	json.NewEncoder(w).Encode(overview)
+}
+
+func getTypeMap(w http.ResponseWriter, r *http.Request) {
+	if app.conv == nil || app.driver == "" {
+		http.Error(w, fmt.Sprintf("Schema is not converted or Driver is not configured properly. Please retry converting the database to spanner."), http.StatusNotFound)
+		return
+	}
+	var typeMap map[string][]typeIssue
+	switch app.driver {
+	case "mysql", "mysqldump":
+		typeMap = mysqlTypeMap
+	case "postgres", "pg_dump":
+		typeMap = postgresTypeMap
+	default:
+		http.Error(w, fmt.Sprintf("Driver : '%s' is not supported", app.driver), http.StatusBadRequest)
+		return
+	}
+	// Filter typeMap so it contains just the types SrcSchema uses.
+	filteredTypeMap := make(map[string][]typeIssue)
+	for _, srcTable := range app.conv.SrcSchema {
+		for _, colDef := range srcTable.ColDefs {
+			if _, ok := filteredTypeMap[colDef.Type.Name]; ok {
+				continue
+			}
+			filteredTypeMap[colDef.Type.Name] = typeMap[colDef.Type.Name]
+		}
+	}
+	w.WriteHeader(http.StatusOK)
+	json.NewEncoder(w).Encode(filteredTypeMap)
+}
+
+func setTypeMapGlobal(w http.ResponseWriter, r *http.Request) {
+	reqBody, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("Body Read Error : %v", err), http.StatusInternalServerError)
+		return
+	}
+	var t map[string]string
+	err = json.Unmarshal(reqBody, &t)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("Request Body parse error : %v", err), http.StatusBadRequest)
+		return
+	}
+
+	for k, v := range app.conv.SpSchema {
```

Comment:
If I understand correctly, this is essentially re-doing the type mapping part of schemaToDDL in mysql/toddl.go and postgres/toddl.go, using the type mapping specified by t, subject to the allowable mapping specified by toSpannerTypeMySQL and toSpannerTypePostgres.

It assumes that schemaToDDL has been called on app.conv (so that app.conv.SpSchema, app.conv.SrcSchema and app.conv.ToSource are all in sync).

There is some subtly here. In particular we need to preserve all previous changes that have been done via the UI e.g. dropped columns, renamed columns, changed primary keys etc. Suggest a comment along the following lines:

// Redo source-to-Spanner type mapping using t (the mapping specified in the http request).
// We drive this process by iterating over the Spanner schema because we want to preserve all
// other customizations that have been performed via the UI (dropping columns, renaming columns
// etc). In particular, note that we can't just blindly redo schema conversion (using an appropriate
// version of 'toDDL' with the new type mapping).

Separately I think we should have a TODO here to fold this functionality into toddl.go (once we've merged mysql and postgres versions of toddl.go!). I think it would be better to have all of this related functionality in one place. And maybe we can find a way to refactor and have avoid some of the duplication -- although I assume a key part of the behaviour of setTypeMapGlobal is that it doesn't change anything else, an in particular that it doesn't redo the entire schema conversion and overwrite dropped columns, dropped foreign keys and other user changes.

How about:

// TODO: Move the type remapping function to toddl.go (once we've merged
// dynamodb/toddl.go,  mysql/toddl.go and postgres/toddl.go).
// Consider some refactoring to reduce code duplication (although note
// that this type remapping has to preserve all previous changes done via the UI!)

Reply:
added first comment as mentioned, but i believe second comment need to be in toSpannerTypePostgres and toSpannerTypeMySQL, so added there.